Add support for List-Unsubscribe

TheLastProject opened the pull request:
Fixes #2943 

This is still a work-in-progress, but I wanted to make this a merge request to get some early feedback as this is my first attempt at working with K-9 Mail's codebase.

There is one main issue with this: for big messages, the List-Unsubscribe header may not be known (related to #5240 and https://github.com/k9mail/k-9/issues/5237#issuecomment-815955904). Currently, the button just isn't shown until the headers are available, which is an awful experience.

What would be the best way to go about this? I see a few options:
1. Always download all headers (increased data usage)
2. Don't show an unsubscribe button in the message but move it into the overflow menu at the top (an show an "This email did not contain any unsubscribing instructions" when clicking on lack of List-Unsubscribe?)

I am also wondering if K-9 Mail already contains some kind of helper module to deal with `<data_we_want>,<data_we_want>` values?

cketti commented:
K-9 Mail supports a way of always fetching specific headers. See `RealImapFolder`.

https://github.com/k9mail/k-9/blob/e619d381437191c87192cd7aad21d449951c9001/mail/protocols/imap/src/main/java/com/fsck/k9/mail/store/imap/RealImapFolder.kt#L539-L543

The most difficult part about a feature like this is figuring out how to integrate it into the UI. Unsubscribing from a mailing list is probably an action that is rarely used. So it shouldn't be displayed prominently. Maybe put it into the toolbar's overflow menu?

TheLastProject commented:
> K-9 Mail supports a way of always fetching specific headers. See `RealImapFolder`.

I'm kinda unsure how this works. Could you explain in more detail how I would ensure the List-Unsubscribe header to always be fetched?

> Maybe put it into the toolbar's overflow menu?

I've hereby done that.

RFC 2369 allows multiple unsubscribe URLs, my code assumes the first sane (http/https/mailto) option is the preferred option. Is that reasonable? Or should we prefer the first mailto option?

Also, I should probably write tests for this. Would testing parsing the header be sufficient testing? I'm not sure how to do UI testing in K-9 Mail's codebase.

cketti commented:
> I'm kinda unsure how this works. Could you explain in more detail how I would ensure the List-Unsubscribe header to always be fetched?

The string added to `fetchFields` is part of the command that will be sent to the IMAP server. The syntax is `BODY.PEEK[HEADER.FIELDS (space separated list of header field names)]`. Just add `list-unsubscribe` to the string:

```kotlin
 fetchFields.add( 
     "BODY.PEEK[HEADER.FIELDS (date subject from content-type to cc " + 
         "reply-to message-id references in-reply-to list-unsubscribe " + 
         K9MailLib.IDENTITY_HEADER + " " + K9MailLib.CHAT_HEADER + ")]" 
 ) 
```

> RFC 2369 allows multiple unsubscribe URLs, my code assumes the first sane (http/https/mailto) option is the preferred option. Is that reasonable? Or should we prefer the first mailto option?

I have no strong preference. As an email client we should probably prefer the mailto option (if there is one). There's a remote chance the app could run on a device without a browser. But we know for sure an email app is installed.

> Also, I should probably write tests for this. Would testing parsing the header be sufficient testing? I'm not sure how to do UI testing in K-9 Mail's codebase.

We currently don't have UI tests. But the header parsing code should be tested. Please put the code in its own class.

The header parsing should also happen when the message is loaded. The result can then be part of `MessageViewInfo`.

TheLastProject commented:
> The unsubscribe button still doesn't show until I click "Download full message".

Actually, it works fine, I just forgot to think about that old emails wouldn't update

cketti commented:
>> The header parsing should also happen when the message is loaded. The result can then be part of `MessageViewInfo`.
>
> I'm not sure where to hook this up, K-9's codebase is very different than what I'm used to, can you give more pointers?

The general idea is that `MessageViewInfo` is a data class that hold all information necessary for the UI to display the message. Instances of `MessageViewInfo` are created by `MessageViewInfoExtractor`. You need to find the right place in this class to call out to your code to extract the unsubscribe URI if there is one, then add it to `MessageViewInfo`.

TheLastProject commented:
I hope this latest commit is the change as you wanted it to be implemented. The method is now called by the MessageViewInfoExtractor class and added to the MessageViewInfo. However, given MessageList has no direct reference to MessageViewInfo, but a reference to MessageViewFragment I hooked it into the onMessageViewInfoLoad* functions.

cketti commented:
The place you found in `MessageViewInfoExtractor` looks good :+1:

---

Instead of putting the code to handle the unsubscribe button into `MessageList`, you could delegate this to `MessageViewFragment`.

```kotlin

override fun onOptionsItemSelected(item: MenuItem): Boolean {
    …
    } else if (id == R.id.unsubscribe) {
        messageViewFragment!!.onUnsubscribe()
        return true
    }
    …
}
```

---

By only returning a `Uri`, part of the "business logic" is duplicated in the UI layer. Instead, consider creating content-specific data types, e.g.

```kotlin
sealed interface UnsubscribeUri {
    val uri: Uri    
}

data class MailtoUnsubscribeUri(override val uri: Uri) : UnsubscribeUri
data class HttpsUnsubscribeUri(override val uri: Uri) : UnsubscribeUri
```

It doesn't make a huge difference because you still need to create different Intents. But doing different things based on the data type is cleaner than having to look at the contents of a Uri.

---

Please have a look at [RFC 2369](https://datatracker.ietf.org/doc/html/rfc2369) for the syntax of the `List-Unsubscribe` header field. The tests seem to create messages using multiple header fields. But that's not how `List-Unsubscribe` is specified or used in practice.

TheLastProject commented:
> Instead of putting the code to handle the unsubscribe button into MessageList, you could delegate this to MessageViewFragment.

Done

> By only returning a Uri, part of the "business logic" is duplicated in the UI layer. Instead, consider creating content-specific data types, e.g.

I think I've implemented this as you asked.

> Please have a look at [RFC 2369](https://datatracker.ietf.org/doc/html/rfc2369) for the syntax of the List-Unsubscribe header field. The tests seem to create messages using multiple header fields. But that's not how List-Unsubscribe is specified or used in practice.

Oops, I totally messed that up when using the ListHeaders test as a base. Good catch, fixed now.

cketti commented on app/core/src/main/java/com/fsck/k9/helper/ListUnsubscribeHelper.kt:
```diff
@@ -0,0 +1,63 @@
+            if (headerValues.isEmpty()) {
+                return null
+            }
+            val listUnsubscribeUris: MutableList<Uri> = ArrayList()
```
This can be written as

```kotlin
val listUnsubscribeUris = mutableListOf<Uri>()
```

cketti commented on app/core/src/main/java/com/fsck/k9/helper/ListUnsubscribeHelper.kt:
```diff
@@ -0,0 +1,63 @@
+            }
+            val listUnsubscribeUris: MutableList<Uri> = ArrayList()
+            for (headerValue in headerValues) {
+                for (headerValuePart in headerValue.split(",")) {
```
Nothing is ever easy. URLs can contain commas. So this will fail to correctly extract e.g. this URL:

```eml
List-Unsubscribe: <https://domain.example/one,two>
```

Not splitting on a comma and changing the regular expressions to match non-greedily should be good enough for us.

```kotlin
private val MAILTO_CONTAINER_PATTERN = Pattern.compile("<(mailto:.+?)>")
private val HTTPS_CONTAINER_PATTERN = Pattern.compile("<(https:.+?)>")
```

cketti commented on app/core/src/test/java/com/fsck/k9/helper/ListUnsubscribeHelperTest.kt:
```diff
@@ -0,0 +1,89 @@
+import org.junit.Assert
+import org.junit.Test
+
+class ListUnsubscribeHelperTest : K9RobolectricTest() {
```
This test only needs Robolectric, but none of the K-9 machinery running. So these tests will run a bit faster if we inherit from `RobolectricTest`.

cketti commented on app/core/src/test/java/com/fsck/k9/helper/ListUnsubscribeHelperTest.kt:
```diff
@@ -0,0 +1,89 @@
+            "<mailto:unsubscribe@example.com>"
+        )
+        val result = ListUnsubscribeHelper.getPreferredListUnsubscribeUri(message)
+        Assert.assertTrue(result is MailtoUnsubscribeUri)
+        Assert.assertEquals("mailto:unsubscribe@example.com", result!!.uri.toString())
```
In new code we're using [Truth](https://truth.dev/) for assertions. Combined with kotlin.test's `assertNotNull` we end up with:

```kotlin
assertNotNull(result) { unsubscribeUri ->
    assertThat(unsubscribeUri).isInstanceOf(MailtoUnsubscribeUri::class.java)
    assertThat(unsubscribeUri.uri.toString()).isEqualTo("mailto:unsubscribe@example.com")
}
```

But because we're using data classes we could also do this:

```kotlin
assertThat(result).isEqualTo(MailtoUnsubscribeUri("mailto:unsubscribe@example.com".toUri()))
```
```

TheLastProject replied:
I went for the first example, because I think the second example doesn't actually test if the correct UnsubscribeUri type is returned, because both versions have toUri().

cketti replied:
They are tests. If you're not sure they're testing the right thing, try to make them fail.

TheLastProject replied:
Ah, yeah, that's my bad. I read your example as:

```
assertThat(result).isEqualTo(MailtoUnsubscribeUri("mailto:unsubscribe@example.com").toUri())
```

instead of

```
assertThat(result).isEqualTo(MailtoUnsubscribeUri("mailto:unsubscribe@example.com".toUri()))
```

Long day at work today :)

cketti commented on app/core/src/test/java/com/fsck/k9/helper/ListUnsubscribeHelperTest.kt:
```diff
@@ -0,0 +1,89 @@
+
+class ListUnsubscribeHelperTest : K9RobolectricTest() {
+    @Test
+    fun getListUnsubscribeUrl_shouldAcceptMailto() {
```
If you want to, you can use slightly easier to read function names. Kotlin allows using spaces and other characters. See e.g.

https://github.com/k9mail/k-9/blob/e1ff582cac7191254dc00d61ebfcab63bfce8011/mail/protocols/smtp/src/test/java/com/fsck/k9/mail/transport/smtp/SmtpResponseParserTest.kt#L623

cketti commented on app/core/src/main/java/com/fsck/k9/helper/ListUnsubscribeHelper.kt:
```diff
@@ -0,0 +1,63 @@
+        // As K-9 Mail is an email client, we prefer a mailto: unsubscribe method
+        // but if none is found, a https URL is acceptable too
+        fun getPreferredListUnsubscribeUri(message: Message?): UnsubscribeUri? {
+            if (message == null) {
+                return null
+            }
```
This function is never called with `message` being `null`. I suggest changing the parameter's type to `Message` and getting rid of this check.

TheLastProject commented:
> You can fix issues reported by ktlint by running `gradlew ktlintFormat`.

Ah nice, I found the command but was trying to find the output. Didn't realize it fixed it automatically :)

TheLastProject commented:
Sorry for the failing test, fixed that just now.

cketti commented:
Thank you :+1: 

Apologies for taking so long to get this merged.

TheLastProject commented:
No need to apologize! Thank you for being so patient with me as I was trying to figure out K-9's codebase and merging it in the end. I really appreciate it :)
